## *Άσκηση 19*

In [7]:
import statistics as stat
import random
random.seed(2020)  # for reproducibility
from simple_markov_chain_lib import markov_chain

p = 1/6
for p in [1/6,1/12]:
    print ("p = ",p);
    # A dictionary for the initial distibution. 
    # We prescribe the initial distribution
    init_probs = {1: 1.0} 
    
    # A dictionary for the transition probability  matrix. 
    # Every state-key corresponds to a list with tuples of (Next_State,Probability) 
    markov_table = {
        1: {2: 1.},
        2: {2: 2/3, 3: 1/3},
        3: {1: p, 2: 1-p}
    }
    
    # Ok... we are ready know
    # Let's construct a Markov Chain. So let's call the constructor
    mc = markov_chain(markov_table, init_probs)
    for init in [1,3]:
        print ("Initial state =",init)
        init_probs = {init: 1.0} 
        for N in [50,200,20000]:
            print ("N = ",N)
            ## Experiment parameters
            #N = 1000     # number of samples
            steps = 20   # the target time
            counter = 0  # to count the number of times the event {X_40  = 1} occurs

            ## Simulation
            for i in range(N):
                mc.start()  # new experiment
                for j in range(steps):  mc.move()
                if mc.running_state == 1:  counter += 1

            phat = counter / N

            print(
                """
                We executed {0} times the first {1} steps of the markov chain
                and we captured the running state in state one {2} times.
                So we estimate the Pr[X_{1} = 1 | X_0 = 1] to be {3}
                """.format(N, steps, counter, phat)
            )
            ## import the library statistics. We will use it to compute the mean and variance of our list
            estimates=[] ## create the empty list estimates.
            for i in range(N): estimates.append(random.gauss(0,1)) 
                ## in each of 100 runs generate a sample from the standard normal distribution and append it to the list estimates
            print(
                """ 
                The sample mean is {0:.5f} and the sample variance is {1:.5f}
                """.format(stat.mean(estimates), stat.variance(estimates))
            )
                ## compute the SAMPLE mean and variance of the elements in our list keeping only 5 decimal digits

p =  0.16666666666666666
Initial state = 1
N =  50

                We executed 50 times the first 20 steps of the markov chain
                and we captured the running state in state one 0 times.
                So we estimate the Pr[X_20 = 1 | X_0 = 1] to be 0.0
                
 
                The sample mean is 0.19169 and the sample variance is 1.13730
                
N =  200

                We executed 200 times the first 20 steps of the markov chain
                and we captured the running state in state one 9 times.
                So we estimate the Pr[X_20 = 1 | X_0 = 1] to be 0.045
                
 
                The sample mean is -0.04632 and the sample variance is 0.84884
                
N =  20000

                We executed 20000 times the first 20 steps of the markov chain
                and we captured the running state in state one 784 times.
                So we estimate the Pr[X_20 = 1 | X_0 = 1] to be 0.0392
                
 
                The

In [16]:
#Some Linear Algebra
import numpy as np
for p in [1/6,1/12]:
  myarray = np.array([[0,1,0],[0,2/3,1/3],[p,1-p,0]])
  for n in range(20): myarray = myarray.dot(myarray)
  print (myarray)

[[0.04 0.72 0.24]
 [0.04 0.72 0.24]
 [0.04 0.72 0.24]]
[[0.02040816 0.73469388 0.24489796]
 [0.02040816 0.73469388 0.24489796]
 [0.02040816 0.73469388 0.24489796]]


## Απαντήσεις

1.  >Για p=1/6 έχουμε: <br>
    >N = 50 -> Pr[X_20 = 1 | X_0 = 1] = 0 <br>
    >N = 200 -> Pr[X_20 = 1 | X_0 = 1] = 0.045 <br>
    >N = 20.000 -> Pr[X_20 = 1 | X_0 = 1] = 0.0392 <br>
    >Για p=1/12 έχουμε: <br>
    >N = 50 -> Pr[X_20 = 1 | X_0 = 1] = 0 <br>
    >N = 200 -> Pr[X_20 = 1 | X_0 = 1] = 0.02 <br>
    >N = 20.000 -> Pr[X_20 = 1 | X_0 = 1] = 0.0209 <br>
2. Σύμφωνα με τον υπολογισμό της θεωρητικής τιμής, ακριβώς από πάνω παρατηρούμε ότι οι θεωρητικές τιμές είναι πάρα πολύ κοντά στις πειραματικές. <br>
3.  >Για p=1/6 έχουμε δειγματική διασπορά: <br>
    >N = 50 -> sample variance = 1.13730 <br>
    >N = 200 -> sample variance = 0.84884 <br>
    >N = 20.000 -> sample variance = 0.99653 <br>
    >Για p=1/12 έχουμε δειγματική διασπορά: <br>
    >N = 50 -> sample variance = 1.28539 <br>
    >N = 200 -> sample variance = 0.97237 <br>
    >N = 20.000 -> sample variance = 0.99169 <br>
4. Παρατηρούμε ότι τα παραπάνω αποτελέσματα είναι πανομοιότυπα.
  
  

## *Άσκηση Τένις*
![alt text](https://www.wolfram.com/mathematica/new-in-9/markov-chains-and-queues/HTMLImages.en/analyze-a-tennis-game/O_23.png)

In [8]:
import statistics as stat
import numpy as np
import random
random.seed(2020)  # for reproducibility
from simple_markov_chain_lib import markov_chain

p = 0.6
q = 1-p
        #   0       1       2       3       4      
States = ["0-0", "15-0", "0-15", "30-0", "15-15",
        #  5       6       7       8       9     
         "0-30", "40-0","30-15","15-30","0-40",
        #   10      11       12     13      14   
         "40-15","Deuce", "15-40","GameA","AdvA",
        #    15      16            
           "AdvB","GameB"]
# A dictionary for the initial distibution. 
# We prescribe the initial distribution
 
 
# A dictionary for the transition probability  matrix. 
# Every state-key corresponds to a list with tuples of (Next_State,Probability) 
markov_table = {
    States[0]: {States[1]: p, States[2]: q},
    States[1]: {States[3]: p, States[4]: q},
    States[2]: {States[4]: p, States[5]: q},
    States[3]: {States[6]: p, States[7]: q},
    States[4]: {States[7]: p, States[8]: q},
    States[5]: {States[8]: p, States[9]: q},
    States[6]: {States[13]: p, States[10]: q},
    States[7]: {States[10]: p, States[11]: q},
    States[8]: {States[11]: p, States[12]: q},
    States[9]: {States[12]: p, States[16]: q},
    States[10]: {States[13]: p, States[14]: q},
    States[11]: {States[14]: p, States[15]: q},
    States[12]: {States[15]: p, States[16]: q},
    States[13]: {States[13]: 1.},
    States[14]: {States[13]: p, States[11]: q},
    States[15]: {States[11]: p, States[16]: q},
    States[16]: {States[16]: 1.}
}
init_probs = {States[0]: 1.0} 
# Ok... we are ready know
# Let"s construct a Markov Chain. So let"s call the constructor
mc = markov_chain(markov_table, init_probs)
for N in [50,200,20000]:
    print ("N = ",N)
    ## Experiment parameters
    #N = 1000     # number of samples
    steps = 0  # the target time
    stepsTrack = []
    counter = 0  # to count the number of times the event {X_40  = 1} occurs

    ## Simulation
    for i in range(N):
        mc.start()  # new experiment
        while(mc.running_state not in [States[13],States[16]]):  
            mc.move()
            steps += 1
        if mc.running_state == States[13]:  
            counter += 1
            stepsTrack.append(steps)
        steps = 0
    phat = counter / N

    print(
        """
        We executed {0} times the first {1} steps until game ends of the markov chain
        and we captured the running state in state {4} {2} times.
        So we estimate the Pr[X_{1} = {4} | X_0 = {5}] to be {3}
        """.format(N, int(np.round(np.mean(stepsTrack))), counter, phat,States[13],States[0])
    )

    ## import the library statistics. We will use it to compute the mean and variance of our list
    estimates=[] ## create the empty list estimates.
    for i in range(N): estimates.append(random.gauss(0,1)) 
        ## in each of 100 runs generate a sample from the standard normal distribution and append it to the list estimates
    print(
        """ 
        The sample mean is {0:.5f} and the sample variance is {1:.5f}
        """.format(stat.mean(estimates), stat.variance(estimates))
    )
        ## compute the SAMPLE mean and variance of the elements in our list keeping only 5 decimal digits

N =  50

        We executed 50 times the first 6 steps until game ends of the markov chain
        and we captured the running state in state GameA 38 times.
        So we estimate the Pr[X_6 = GameA | X_0 = 0-0] to be 0.76
        
 
        The sample mean is -0.03709 and the sample variance is 0.84798
        
N =  200

        We executed 200 times the first 7 steps until game ends of the markov chain
        and we captured the running state in state GameA 143 times.
        So we estimate the Pr[X_7 = GameA | X_0 = 0-0] to be 0.715
        
 
        The sample mean is 0.02324 and the sample variance is 0.92712
        
N =  20000

        We executed 20000 times the first 6 steps until game ends of the markov chain
        and we captured the running state in state GameA 14708 times.
        So we estimate the Pr[X_6 = GameA | X_0 = 0-0] to be 0.7354
        
 
        The sample mean is 0.00069 and the sample variance is 1.00451
        


# Τενις
Παραπάνω μοντελοποιήσαμε τις πιθανές εκβάσεις στο πρωτάθλημα τένις με παρόμοιο τρόπο.

## Αποτελέσματα
Παρατηρούμε ότι αν η πιθανότητα ο παίκτης Α να κερδίσει τον πόντο είναι p = 0.6 τότε η πιθανότητα να κερδίσει ο παίκτης Α το παιχνίδι είναι περίπου 73%. Επίσης υπολογίσαμε ότι η μέση διάρκεια ενός παιχνιδιού είναι 6-7 γύρους ως που να βρεθεί κάποιος νικητής.